File: npm/corsa_oxlint/ts/rules/await_thenable.ts
import { isIdentifierNamed, memberObject, memberPropertyName, stripChainExpression } from "./ast";
import { createNativeRule } from "./rule_creator";
import { isPromiseLikeNode } from "./type_utils";

export const awaitThenableRule = createNativeRule(
  "await-thenable",
  {
    docs: {
      description: "Disallow awaiting non-thenable values.",
    },
    messages: {
      unexpected: "Unexpected await of a non-thenable value.",
    },
  },
  (context) => ({
    AwaitExpression(node: any) {
      if (!isPromiseLikeNode(context, node.argument) && !isObviouslyPromiseLike(node.argument)) {
        context.report({ node, messageId: "unexpected" });
      }
    },
  }),
);

function isObviouslyPromiseLike(node: any): boolean {
  const current = stripChainExpression(node);
  if (current?.type === "NewExpression" && isIdentifierNamed(current.callee, "Promise")) {
    return true;
  }
  if (current?.type !== "CallExpression") {
    return false;
  }
  return (
    memberPropertyName(current.callee) === "resolve" &&
    isIdentifierNamed(memberObject(current.callee), "Promise")
  );
}

File: src/core/corsa_core/src/lint/rules/await_thenable.rs
use super::super::{LintNode, RuleContext, RuleMessage, RustLintRule};
use crate::lint::helpers::{is_obviously_promise_like, is_promise_like_node};

/// Type-aware rule that rejects `await` on values without a thenable shape.
#[derive(Clone, Copy, Debug, Default)]
pub struct AwaitThenableRule;

const MESSAGES: &[RuleMessage] = &[RuleMessage {
    id: "unexpected",
    description: "Unexpected await of a non-thenable value.",
}];
const LISTENERS: &[&str] = &["AwaitExpression"];

impl RustLintRule for AwaitThenableRule {
    fn name(&self) -> &'static str {
        "await-thenable"
    }

    fn docs_description(&self) -> &'static str {
        "Disallow awaiting non-thenable values."
    }

    fn messages(&self) -> &'static [RuleMessage] {
        MESSAGES
    }

    fn listeners(&self) -> &'static [&'static str] {
        LISTENERS
    }

    fn check(&self, ctx: &mut RuleContext<'_>, node: &LintNode) {
        if node.kind != "AwaitExpression" {
            return;
        }
        let Some(argument) = node.child("argument") else {
            return;
        };
        if !is_promise_like_node(argument) && !is_obviously_promise_like(argument) {
            ctx.report("unexpected", node.range);
        }
    }
}
